crates/extensions/c8y_firmware_manager/src/tests.rs

```diff
@@ -0,0 +1,787 @@
+                &Topic::new_unchecked(C8Y_CHILD_PUBLISH_TOPIC_NAME),
+                format!("502,c8y_Firmware,\"{expected_failure_text}\"\n"),
+            ),
+        ])
+        .await;
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+async fn handle_child_response_while_busy_downloading() -> Result<(), DynError> {
+    let mut ttd = TempTedgeDir::new();
+
+    let (
+        mut mqtt_message_box,
+        mut _jwt_message_box,
+        mut _timer_message_box,
+        mut _downloader_message_box,
+    ) = spawn_firmware_manager(&mut ttd, DEFAULT_REQUEST_TIMEOUT_SEC, true, true).await?;
+
+    // Ignore SmartREST 500.
+    mqtt_message_box.skip(1).await;
+
+    // Publish a firmware update operation to child device that requires new file download.
+    let c8y_firmware_update_message1 = MqttMessage::new(
+        &Topic::new_unchecked("c8y/s/ds"),
+        format!("515,child-2,{FIRMWARE_NAME},{FIRMWARE_VERSION},http://firmware2"),
+    );
+    mqtt_message_box.send(c8y_firmware_update_message1).await?;
+
+    // Publish a firmware update operation to another child device that does NOT need new file download.
+    publish_smartrest_firmware_operation(&mut mqtt_message_box).await?;
+
+    // This firmware update request is for the one that doesn't need a new file download.
+    let firmware_update_request = mqtt_message_box.recv().await.unwrap();
+    let operation_id = get_operation_id_from_firmware_update_request(firmware_update_request);
+
+    // Publish a successful RESPONSE while downloading is in progress for another child device.
+    publish_firmware_update_response("successful", &operation_id, &mut mqtt_message_box).await?;
+
+    // Assert EXECUTING SmartREST message, installed firmware SmartREST message, SUCCESSFUL SmartREST message
+    mqtt_message_box
+        .assert_received([
+            MqttMessage::new(
+                &Topic::new_unchecked(C8Y_CHILD_PUBLISH_TOPIC_NAME),
+                "501,c8y_Firmware\n",
+            ),
+            MqttMessage::new(
+                &Topic::new_unchecked(C8Y_CHILD_PUBLISH_TOPIC_NAME),
+                format!("115,{FIRMWARE_NAME},{FIRMWARE_VERSION},{DOWNLOAD_URL}"),
+            ),
+            MqttMessage::new(
+                &Topic::new_unchecked(C8Y_CHILD_PUBLISH_TOPIC_NAME),
+                "503,c8y_Firmware,\n",
+            ),
+        ])
+        .await;
+
+    Ok(())
+}
+
+fn get_operation_id_from_firmware_update_request(mqtt_message: MqttMessage) -> String {
+    serde_json::from_str::<serde_json::Value>(mqtt_message.payload.as_str().unwrap())
+        .expect("Deserialize JSON")
+        .get("id")
+        .expect("'id' field exists")
+        .as_str() // Cannot use "to_string()" directly as it will include `\`.
+        .expect("string")
+        .to_owned()
+}
+
+async fn publish_smartrest_firmware_operation(
+    mqtt_message_box: &mut TimedMessageBox<SimpleMessageBox<MqttMessage, MqttMessage>>,
+) -> Result<(), DynError> {
```

Comment:
Maybe this kind of function could be wrapped around structs that represents actual entities as the c8y or child devices.

So the tests could be written as follows:

```
c8y.publish_smartrest_firmware_operation().await?;
let operation_id = child_device.receive_firmware_operation().await?;
child_device.publish_firmware_update_response("successful", &operation_id).await?;
c8y.receive_firmware_update_response("successful").await?;
```